File: go/encode.go
package msgpackblob

import "math"

func encNil(out *[]byte) { *out = append(*out, mpNil) }

func encBool(out *[]byte, v bool) {
	if v {
		*out = append(*out, mpTrue)
	} else {
		*out = append(*out, mpFalse)
	}
}

func encInteger(out *[]byte, x int64) {
	if x >= 0 {
		switch {
		case x <= 0x7f:
			*out = append(*out, byte(x))
		case x <= 0xff:
			*out = append(*out, mpUint8, byte(x))
		case x <= 0xffff:
			*out = append(*out, mpUint16)
			put16(out, uint16(x))
		case x <= 0xffffffff:
			*out = append(*out, mpUint32)
			put32(out, uint32(x))
		default:
			*out = append(*out, mpUint64)
			put64(out, uint64(x))
		}
	} else {
		switch {
		case x >= -32:
			*out = append(*out, byte(x))
		case x >= -128:
			*out = append(*out, mpInt8, byte(x))
		case x >= -32768:
			*out = append(*out, mpInt16)
			put16(out, uint16(x))
		case x >= -2147483648:
			*out = append(*out, mpInt32)
			put32(out, uint32(x))
		default:
			*out = append(*out, mpInt64)
			put64(out, uint64(x))
		}
	}
}

func encUnsigned(out *[]byte, x uint64) {
	switch {
	case x <= 0x7f:
		*out = append(*out, byte(x))
	case x <= 0xff:
		*out = append(*out, mpUint8, byte(x))
	case x <= 0xffff:
		*out = append(*out, mpUint16)
		put16(out, uint16(x))
	case x <= 0xffffffff:
		*out = append(*out, mpUint32)
		put32(out, uint32(x))
	default:
		*out = append(*out, mpUint64)
		put64(out, x)
	}
}

func encReal(out *[]byte, d float64) {
	*out = append(*out, mpFloat64)
	put64(out, math.Float64bits(d))
}

func encReal32(out *[]byte, f float32) {
	*out = append(*out, mpFloat32)
	put32(out, math.Float32bits(f))
}

func encString(out *[]byte, s []byte) {
	n := len(s)
	switch {
	case n <= 31:
		*out = append(*out, byte(mpFixstrMask|n))
	case n <= 0xff:
		*out = append(*out, mpStr8, byte(n))
	case n <= 0xffff:
		*out = append(*out, mpStr16)
		put16(out, uint16(n))
	default:
		*out = append(*out, mpStr32)
		put32(out, uint32(n))
	}
	*out = append(*out, s...)
}

func encBinary(out *[]byte, data []byte) {
	n := len(data)
	switch {
	case n <= 0xff:
		*out = append(*out, mpBin8, byte(n))
	case n <= 0xffff:
		*out = append(*out, mpBin16)
		put16(out, uint16(n))
	default:
		*out = append(*out, mpBin32)
		put32(out, uint32(n))
	}
	*out = append(*out, data...)
}

func encExt(out *[]byte, typeCode int8, data []byte) {
	n := len(data)
	switch n {
	case 1:
		*out = append(*out, mpFixext1)
	case 2:
		*out = append(*out, mpFixext2)
	case 4:
		*out = append(*out, mpFixext4)
	case 8:
		*out = append(*out, mpFixext8)
	case 16:
		*out = append(*out, mpFixext16)
	default:
		switch {
		case n <= 0xff:
			*out = append(*out, mpExt8, byte(n))
		case n <= 0xffff:
			*out = append(*out, mpExt16)
			put16(out, uint16(n))
		default:
			*out = append(*out, mpExt32)
			put32(out, uint32(n))
		}
	}
	*out = append(*out, byte(typeCode))
	*out = append(*out, data...)
}

func encInt8(out *[]byte, x int64)    { *out = append(*out, mpInt8, byte(x)) }
func encInt16(out *[]byte, x int64)   { *out = append(*out, mpInt16); put16(out, uint16(x)) }
func encInt32(out *[]byte, x int64)   { *out = append(*out, mpInt32); put32(out, uint32(x)) }
func encInt64(out *[]byte, x int64)   { *out = append(*out, mpInt64); put64(out, uint64(x)) }
func encUint8(out *[]byte, x uint64)  { *out = append(*out, mpUint8, byte(x)) }
func encUint16(out *[]byte, x uint64) { *out = append(*out, mpUint16); put16(out, uint16(x)) }
func encUint32(out *[]byte, x uint64) { *out = append(*out, mpUint32); put32(out, uint32(x)) }
func encUint64(out *[]byte, x uint64) { *out = append(*out, mpUint64); put64(out, x) }

func encArrayHeader(out *[]byte, count uint32) {
	switch {
	case count <= 15:
		*out = append(*out, byte(mpFixarrayMask|count))
	case count <= 0xffff:
		*out = append(*out, mpArray16)
		put16(out, uint16(count))
	default:
		*out = append(*out, mpArray32)
		put32(out, count)
	}
}

func encMapHeader(out *[]byte, count uint32) {
	switch {
	case count <= 15:
		*out = append(*out, byte(mpFixmapMask|count))
	case count <= 0xffff:
		*out = append(*out, mpMap16)
		put16(out, uint16(count))
	default:
		*out = append(*out, mpMap32)
		put32(out, count)
	}
}

func encTimestamp(out *[]byte, sec int64, nsec uint32) {
	if nsec == 0 && sec >= 0 && sec <= 0xffffffff {
		*out = append(*out, mpFixext4, 0xff)
		put32(out, uint32(sec))
	} else if sec >= 0 && sec <= 0x3FFFFFFFF {
		*out = append(*out, mpFixext8, 0xff)
		put64(out, (uint64(nsec)<<34)|uint64(sec))
	} else {
		*out = append(*out, mpExt8, 12, 0xff)
		put32(out, nsec)
		put64(out, uint64(sec))
	}
}

func encodeValue(out *[]byte, v Value) {
	switch v.ty {
	case TypeNil:
		encNil(out)
	case TypeTrue:
		encBool(out, true)
	case TypeFalse:
		encBool(out, false)
	case TypeInteger:
		switch v.intWidth {
		case WidthInt8:
			encInt8(out, v.AsInt64())
		case WidthInt16:
			encInt16(out, v.AsInt64())
		case WidthInt32:
			encInt32(out, v.AsInt64())
		case WidthInt64:
			encInt64(out, v.AsInt64())
		case WidthUint8:
			encUint8(out, v.AsUint64())
		case WidthUint16:
			encUint16(out, v.AsUint64())
		case WidthUint32:
			encUint32(out, v.AsUint64())
		case WidthUint64:
			encUint64(out, v.AsUint64())
		default:
			encInteger(out, v.AsInt64())
		}
	case TypeReal:
		encReal(out, v.AsFloat64())
	case TypeFloat32:
		encReal32(out, v.AsFloat32())
	case TypeString:
		encString(out, v.AsBytes())
	case TypeBinary:
		encBinary(out, v.BlobData())
	case TypeExt:
		encExt(out, v.ExtType(), v.BlobData())
	case TypeTimestamp:
		encTimestamp(out, v.TimestampSeconds(), v.TimestampNanoseconds())
	default:
		encNil(out)
	}
}

File: python/msgpack_blob/_encode.py
"""Internal: encoding primitives (mirrors msgpack_blob_encode.cpp).

All functions append to a ``bytearray`` and are shared by the Builder, the JSON
parser and the mutation engine.
"""

from __future__ import annotations

import struct

from . import _format as F
from .value import IntWidth, Type, Value

__all__ = [
    "enc_nil",
    "enc_bool",
    "enc_integer",
    "enc_unsigned",
    "enc_real",
    "enc_real32",
    "enc_string",
    "enc_binary",
    "enc_ext",
    "enc_int8", "enc_int16", "enc_int32", "enc_int64",
    "enc_uint8", "enc_uint16", "enc_uint32", "enc_uint64",
    "enc_array_header",
    "enc_map_header",
    "enc_timestamp",
    "encode_value",
]

_U64 = 0xFFFFFFFFFFFFFFFF


def enc_nil(out: bytearray) -> None:
    out.append(F.MP_NIL)


def enc_bool(out: bytearray, v: bool) -> None:
    out.append(F.MP_TRUE if v else F.MP_FALSE)


def enc_integer(out: bytearray, x: int) -> None:
    """Compact signed-integer encoding (fixint → int64/uint64)."""
    if x >= 0:
        if x <= 0x7F:
            out.append(x)
        elif x <= 0xFF:
            out += bytes((F.MP_UINT8, x))
        elif x <= 0xFFFF:
            out += bytes((F.MP_UINT16,)) + F.w16(x)
        elif x <= 0xFFFFFFFF:
            out += bytes((F.MP_UINT32,)) + F.w32(x)
        else:
            out += bytes((F.MP_UINT64,)) + F.w64(x)
    else:
        if x >= -32:
            out.append(x & 0xFF)
        elif x >= -128:
            out += bytes((F.MP_INT8, x & 0xFF))
        elif x >= -32768:
            out += bytes((F.MP_INT16,)) + F.w16(x)
        elif x >= -2147483648:
            out += bytes((F.MP_INT32,)) + F.w32(x)
        else:
            out += bytes((F.MP_INT64,)) + F.w64(x)


def enc_unsigned(out: bytearray, x: int) -> None:
    """Compact unsigned-integer encoding."""
    x &= _U64
    if x <= 0x7F:
        out.append(x)
    elif x <= 0xFF:
        out += bytes((F.MP_UINT8, x))
    elif x <= 0xFFFF:
        out += bytes((F.MP_UINT16,)) + F.w16(x)
    elif x <= 0xFFFFFFFF:
        out += bytes((F.MP_UINT32,)) + F.w32(x)
    else:
        out += bytes((F.MP_UINT64,)) + F.w64(x)


def enc_real(out: bytearray, d: float) -> None:
    out += bytes((F.MP_FLOAT64,)) + struct.pack(">d", d)


def enc_real32(out: bytearray, f: float) -> None:
    out += bytes((F.MP_FLOAT32,)) + struct.pack(">f", f)


def enc_string(out: bytearray, s: bytes) -> None:
    n = len(s)
    if n <= 31:
        out.append(F.MP_FIXSTR_MASK | n)
    elif n <= 0xFF:
        out += bytes((F.MP_STR8, n))
    elif n <= 0xFFFF:
        out += bytes((F.MP_STR16,)) + F.w16(n)
    else:
        out += bytes((F.MP_STR32,)) + F.w32(n)
    out += s


def enc_binary(out: bytearray, data: bytes) -> None:
    n = len(data)
    if n <= 0xFF:
        out += bytes((F.MP_BIN8, n))
    elif n <= 0xFFFF:
        out += bytes((F.MP_BIN16,)) + F.w16(n)
    else:
        out += bytes((F.MP_BIN32,)) + F.w32(n)
    out += data


def enc_ext(out: bytearray, type_code: int, data: bytes) -> None:
    n = len(data)
    if n == 1:
        out.append(F.MP_FIXEXT1)
    elif n == 2:
        out.append(F.MP_FIXEXT2)
    elif n == 4:
        out.append(F.MP_FIXEXT4)
    elif n == 8:
        out.append(F.MP_FIXEXT8)
    elif n == 16:
        out.append(F.MP_FIXEXT16)
    elif n <= 0xFF:
        out += bytes((F.MP_EXT8, n))
    elif n <= 0xFFFF:
        out += bytes((F.MP_EXT16,)) + F.w16(n)
    else:
        out += bytes((F.MP_EXT32,)) + F.w32(n)
    out.append(type_code & 0xFF)
    out += data


def enc_int8(out: bytearray, x: int) -> None:
    out += bytes((F.MP_INT8, x & 0xFF))


def enc_int16(out: bytearray, x: int) -> None:
    out += bytes((F.MP_INT16,)) + F.w16(x)


def enc_int32(out: bytearray, x: int) -> None:
    out += bytes((F.MP_INT32,)) + F.w32(x)


def enc_int64(out: bytearray, x: int) -> None:
    out += bytes((F.MP_INT64,)) + F.w64(x)


def enc_uint8(out: bytearray, x: int) -> None:
    out += bytes((F.MP_UINT8, x & 0xFF))


def enc_uint16(out: bytearray, x: int) -> None:
    out += bytes((F.MP_UINT16,)) + F.w16(x)


def enc_uint32(out: bytearray, x: int) -> None:
    out += bytes((F.MP_UINT32,)) + F.w32(x)


def enc_uint64(out: bytearray, x: int) -> None:
    out += bytes((F.MP_UINT64,)) + F.w64(x)


def enc_array_header(out: bytearray, count: int) -> None:
    if count <= 15:
        out.append(F.MP_FIXARRAY_MASK | count)
    elif count <= 0xFFFF:
        out += bytes((F.MP_ARRAY16,)) + F.w16(count)
    else:
        out += bytes((F.MP_ARRAY32,)) + F.w32(count)


def enc_map_header(out: bytearray, count: int) -> None:
    if count <= 15:
        out.append(F.MP_FIXMAP_MASK | count)
    elif count <= 0xFFFF:
        out += bytes((F.MP_MAP16,)) + F.w16(count)
    else:
        out += bytes((F.MP_MAP32,)) + F.w32(count)


def enc_timestamp(out: bytearray, sec: int, nsec: int = 0) -> None:
    if nsec == 0 and 0 <= sec <= 0xFFFFFFFF:
        out += bytes((F.MP_FIXEXT4, 0xFF)) + F.w32(sec)
    elif 0 <= sec <= 0x3FFFFFFFF:
        out += bytes((F.MP_FIXEXT8, 0xFF)) + F.w64((nsec << 34) | sec)
    else:
        out += bytes((F.MP_EXT8, 12, 0xFF)) + F.w32(nsec) + F.w64(sec)


def encode_value(out: bytearray, v: Value) -> None:
    """Encode a Value, honouring its integer-width hint."""
    t = v.type()
    if t is Type.NIL:
        enc_nil(out)
    elif t is Type.TRUE:
        enc_bool(out, True)
    elif t is Type.FALSE:
        enc_bool(out, False)
    elif t is Type.INTEGER:
        w = v.int_width()
        if w is IntWidth.INT8:
            enc_int8(out, v.as_int64())
        elif w is IntWidth.INT16:
            enc_int16(out, v.as_int64())
        elif w is IntWidth.INT32:
            enc_int32(out, v.as_int64())
        elif w is IntWidth.INT64:
            enc_int64(out, v.as_int64())
        elif w is IntWidth.UINT8:
            enc_uint8(out, v.as_uint64())
        elif w is IntWidth.UINT16:
            enc_uint16(out, v.as_uint64())
        elif w is IntWidth.UINT32:
            enc_uint32(out, v.as_uint64())
        elif w is IntWidth.UINT64:
            enc_uint64(out, v.as_uint64())
        else:
            enc_integer(out, v.as_int64())
    elif t is Type.REAL:
        enc_real(out, v.as_double())
    elif t is Type.FLOAT32:
        enc_real32(out, v.as_float())
    elif t is Type.STRING:
        enc_string(out, v.as_bytes())
    elif t is Type.BINARY:
        enc_binary(out, v.blob_data())
    elif t is Type.EXT:
        enc_ext(out, v.ext_type(), v.blob_data())
    elif t is Type.TIMESTAMP:
        enc_timestamp(out, v.timestamp_seconds(), v.timestamp_nanoseconds())
    else:
        enc_nil(out)
